server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Keywords.java

```diff
@@ -0,0 +1,167 @@
+
+public class Keywords {
+    public interface KeywordsValidator {
+        void validate(Set<Keyword> keywords);
+    }
+
+    private FlagsBuilder combiner(FlagsBuilder firstBuilder, FlagsBuilder secondBuilder) {
+        return firstBuilder.add(secondBuilder.build());
+    }
+
+    private FlagsBuilder accumulator(FlagsBuilder accumulator, Keyword keyword) {
+        return accumulator.add(keyword.asFlags());
+    }
+
+    public static class KeywordsFactory {
+        private KeywordsValidator validator;
+        private Predicate<Keyword> filter;
+
+        private KeywordsFactory() {
+            validator = keywords -> {};
+            filter = keyword -> true;
+        }
+
+        public KeywordsFactory throwOnAddDraft() {
+            validator = keywords -> Preconditions.checkArgument(
+                !keywords.contains(Keyword.DRAFT), "Cannot add or remove draft flag");
+            return this;
+        }
+
+        public KeywordsFactory throwOnRemoveDraft() {
+            validator = keywords -> Preconditions.checkArgument(
+                keywords.contains(Keyword.DRAFT), "Cannot add or remove draft flag");
+            return this;
+        }
+
+        public KeywordsFactory throwWhenUnsupportedKeyword() {
+            validator = keywords -> Preconditions.checkArgument(
+                keywords.stream().allMatch(Keyword::isSupported), "Does not allow to update 'Deleted' or 'Recent' flag");
+            return this;
+        }
+
+        public KeywordsFactory filterUnsupportedFlags() {
+            filter = keyword -> keyword.isSupported();
+            return this;
+        }
+
+        public Keywords fromMap(Map<String, Boolean> mapKeywords) {
+            Preconditions.checkArgument(mapKeywords.values()
+                .stream()
+                .noneMatch(keywordValue -> !keywordValue), "Keyword must be true");
+            Set<Keyword> setKeywords = mapKeywords.keySet().stream()
+                .map(Keyword::new)
+                .filter(filter)
+                .collect(Guavate.toImmutableSet());
+
+            validator.validate(setKeywords);
+
+            return new Keywords(setKeywords);
+        }
+
+        public Keywords fromOldKeywords(OldKeyword oldKeyword) {
+            ImmutableSet.Builder<Keyword> builder = ImmutableSet.builder();
+            if (oldKeyword.isAnswered().isPresent() && oldKeyword.isAnswered().get()) {
+                builder.add(Keyword.ANSWERED);
+            }
+            if (oldKeyword.isDraft().isPresent() && oldKeyword.isDraft().get()) {
+                builder.add(Keyword.DRAFT);
+            }
+            if (oldKeyword.isFlagged().isPresent() && oldKeyword.isFlagged().get()) {
+                builder.add(Keyword.FLAGGED);
+            }
+            if (oldKeyword.isUnread().isPresent() && !oldKeyword.isUnread().get()) {
+                builder.add(Keyword.SEEN);
+            }
+            return new Keywords(builder.build());
```

Comment:
Idem, reuse **fromSet** here.

(You want to enforce your policies on any input)